### ISSUE TITLE
Print a warning if item gets unregistered

### ISSUE THREAD

**Issue by adrido** [minetest/minetest#8496]
##### Issue type

 - Feature request
 
##### Minetest version

```
5.0.*
```

##### Summary
There should be a warning (including modname, file and linenumber) printed if a item gets unregistered via minetest.unregister_item

Reason: 
Imagine 3 mods. Mod A, B and C. B and C depends on A. C requires a special item of A either for a craft reciepe or just the definition. But mod B is loaded before C and unregisteres an Item of A for some reason. --> This leads to either unexpected results of mod C or may even crash it if it uses the item definition. If you now want to figure out why C is not working right you find out it requires a Item of A. You may check for updates for A or take a look into the sourcecode and only find out that item is registered properly. So A and C are working correct, but where is the failure?

**Comment by ClobberXD** [comment]
Wouldn't a callback be more useful for that purpose?

**Comment by adrido** [comment]
If a mod deletes an essential item of another mod its already too late. How would a callback help there?

**Comment by ClobberXD** [comment]
Printing a warning just lets the modder or server owner know that an item has been unregistered. But a callback can allow much more - let's take your own example: Adding a callback to C can allow it to skip over the parts that depend on the item B unregistered. The warning itself can be printed using this callback if a mod necessitates it. Since this warning has a very specific usecase, adding the warning to the engine is not a good idea.

TL;DR: Callback offers a huge measure of flexibility, compared to a warning being printed, as mods can detect and act upon the former.

**Comment by DS-Minetest** [comment]
In this special example a modder would perhaps not even get the idea that `core.unregister_item` is the problem. Ergo a message might make sense.

But maybe make the message verbose?

**Comment by adrido** [comment]
> Printing a warning just lets the modder or server owner know that an item has been unregistered.

Thats exactly the idea. Do you know how much time it takes to find out why e.g. default:ladder_wood degenerated to a unknown node? And then find out why and which mod caused that.


> Adding a callback to mod C can allow it to skip over the parts that depend on the mod A's item which mod B unregistered. The warning itself can be printed using this callback if a mod necessitates it. Also, since this warning has a very specific use-case, adding it to the engine is not a good idea.

This would requiring heavy changes to all mods to maintain a list of all items from other mods and may check if they exist or not if one particualar mod is installed. Also please don't understand me wrong, *mod A and C working correctly*, the failure is in mod B that breaks A and C. And thats why a warning schould printed, because B can break correctly working mods.



> But maybe make the message verbose?

That would also ok imo, but verbose messages are not printed by default. 

You could compare the warning message with the OS dialog "Do you really want to delete this file". Deleting a file might cause problems, same as unregistering a item. In this case the File(item) is not even deleted by a human(who could recover the File(item), it is deleted automatically by a script.

**Comment by ClobberXD** [comment]
Printing warnings are still possible, if callbacks are implemented. What I'm saying is that if we're going to do something, we might as well do it properly. Since callbacks are functions, mods can do anything upon a trigger, incl. printing warnings.

Ideally, the engine should offer the maximum flexibility possible (within limits™️) by providing "interfaces". It's up to the mods to use these interfaces to implement specific functionality, like printing warnings in this case.

**Comment by SmallJoker** [comment]
`minetest.unregister_item` properly removes the item definition as long it's used before the server initializes all items. It would make less sense to print a warning if the item removal is entirely intentional. If the mods C and B are incompatible, the authors should fix that by themselves.
